File: 740-delete-and-earn/740-delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        nums.sort()
        count = Counter(nums)
        keys = list(count.keys())
        dp = [0] * len(keys)
        
        for i in range(len(keys)):
            if keys[i-1] + 1 == keys[i]:
                dp[i] = max(dp[i-1], (keys[i] * count[keys[i]]) + dp[i-2])
            else:
                dp[i] = max(dp[i-1], (keys[i] * count[keys[i]]) + dp[i-1])
                
        return dp[-1]

File: 740-delete-and-earn/740-delete-and-earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int n = 10001;
        vector<int> dp(n + 1);
        
        for (auto n: nums)
            dp[n] += n;
        
        for (int i = 2; i <= n; i++)
            dp[i] = max(dp[i - 1], dp[i - 2] + dp[i]);
        
        return dp[n];
    }
};